File: debug_height_calc.py
# 高さ計算の詳細デバッグ版
STAGE_UNIT_HEIGHT = 1900
FIRST_LAYER_MIN_HEIGHT_THRESHOLD = 950
TIE_COLUMN_REDUCTION_LARGE = 475
TIE_COLUMN_REDUCTION_SMALL = 130
TIE_COLUMN_MIN_HEIGHT_FOR_LARGE_REDUCTION = 550
TIE_COLUMN_MIN_HEIGHT_FOR_SMALL_REDUCTION = 150

def debug_height_calculation(standard_height, roof_shape, tie_column, railing_count):
    print(f"=== 高さ計算デバッグ ===")
    print(f"入力: height={standard_height}, roof_shape={roof_shape}, tie_column={tie_column}, railing_count={railing_count}")
    
    # Step 1: 基本値
    base_unit_map = {"フラット": 1700, "勾配軒": 1900, "陸屋根": 1800}
    base_unit = base_unit_map.get(roof_shape, 1700)
    remainder = standard_height - base_unit
    stage_unit = STAGE_UNIT_HEIGHT
    
    print(f"Step 1: base_unit={base_unit}, remainder={remainder}, stage_unit={stage_unit}")
    
    # Step 2: 初期段数計算
    initial_stages = 1 + (remainder // stage_unit if remainder > 0 else 0)
    initial_leftover = remainder - (initial_stages - 1) * stage_unit
    
    print(f"Step 2: initial_stages={initial_stages}, initial_leftover={initial_leftover}")
    
    # Step 3: 1層目高さ決定
    if initial_leftover < FIRST_LAYER_MIN_HEIGHT_THRESHOLD:
        first_layer_height = initial_leftover + stage_unit
        print(f"Step 3: initial_leftover({initial_leftover}) < threshold({FIRST_LAYER_MIN_HEIGHT_THRESHOLD}) → first_layer_height = {initial_leftover} + {stage_unit} = {first_layer_height}")
    else:
        first_layer_height = initial_leftover
        print(f"Step 3: initial_leftover({initial_leftover}) >= threshold({FIRST_LAYER_MIN_HEIGHT_THRESHOLD}) → first_layer_height = {first_layer_height}")
    
    # Step 4: 最終段数計算
    remaining_after_first = standard_height - base_unit - first_layer_height
    num_stages = 1 + (remaining_after_first // stage_unit if remaining_after_first > 0 else 0)
    leftover = standard_height - base_unit - (num_stages - 1) * stage_unit
    
    print(f"Step 4: remaining_after_first={remaining_after_first}, num_stages={num_stages}, leftover={leftover}")
    
    # Step 5: ジャッキアップ高さ計算
    jack_up_height = leftover
    reduction_loops = 0
    tie_possible = True
    
    print(f"Step 5 初期: jack_up_height={jack_up_height}")
    
    if tie_column:
        print("根がらみ支柱使用")
        if jack_up_height >= TIE_COLUMN_MIN_HEIGHT_FOR_LARGE_REDUCTION:
            while jack_up_height >= TIE_COLUMN_MIN_HEIGHT_FOR_LARGE_REDUCTION:
                print(f"  大減算: {jack_up_height} - {TIE_COLUMN_REDUCTION_LARGE} = {jack_up_height - TIE_COLUMN_REDUCTION_LARGE}")
                jack_up_height -= TIE_COLUMN_REDUCTION_LARGE
                reduction_loops += 1
            if jack_up_height >= TIE_COLUMN_MIN_HEIGHT_FOR_SMALL_REDUCTION:
                print(f"  小減算: {jack_up_height} - {TIE_COLUMN_REDUCTION_SMALL} = {jack_up_height - TIE_COLUMN_REDUCTION_SMALL}")
                jack_up_height -= TIE_COLUMN_REDUCTION_SMALL
            else:
                tie_possible = False
                print(f"  小減算不可 → tie_possible=False, jack_up_height={jack_up_height}, reduction_loops={reduction_loops}")
        elif jack_up_height >= TIE_COLUMN_MIN_HEIGHT_FOR_SMALL_REDUCTION:
            print(f"  小減算のみ: {jack_up_height} - {TIE_COLUMN_REDUCTION_SMALL} = {jack_up_height - TIE_COLUMN_REDUCTION_SMALL}")
            jack_up_height -= TIE_COLUMN_REDUCTION_SMALL
        else:
            tie_possible = False
            jack_up_height = leftover
            print(f"  減算不可 → tie_possible=False, jack_up_height={jack_up_height}")
    else:
        print("根がらみ支柱不使用")
        while jack_up_height >= TIE_COLUMN_REDUCTION_LARGE:
            print(f"  大減算: {jack_up_height} - {TIE_COLUMN_REDUCTION_LARGE} = {jack_up_height - TIE_COLUMN_REDUCTION_LARGE}")
            jack_up_height -= TIE_COLUMN_REDUCTION_LARGE
            reduction_loops += 1
    
    # Step 6: コマ数計算
    modules_count = 4 + (num_stages - 1) * 4 + reduction_loops
    if railing_count == 3:
        modules_count += 2
    elif railing_count == 2:
        modules_count += 1
    
    print(f"Step 6: modules_count = 4 + ({num_stages} - 1) * 4 + {reduction_loops} + railing_bonus = {modules_count}")
    
    # Step 7: 最終調整（問題の可能性がある箇所）
    final_leftover_for_first_layer = standard_height - base_unit - (num_stages - 1) * stage_unit
    print(f"Step 7 最終調整: final_leftover_for_first_layer = {standard_height} - {base_unit} - {(num_stages - 1) * stage_unit} = {final_leftover_for_first_layer}")
    
    if final_leftover_for_first_layer < FIRST_LAYER_MIN_HEIGHT_THRESHOLD:
        new_first_layer_height = final_leftover_for_first_layer + stage_unit
        print(f"Step 7: {final_leftover_for_first_layer} < {FIRST_LAYER_MIN_HEIGHT_THRESHOLD} → first_layer_height = {new_first_layer_height}")
        first_layer_height = new_first_layer_height
    else:
        print(f"Step 7: {final_leftover_for_first_layer} >= {FIRST_LAYER_MIN_HEIGHT_THRESHOLD} → first_layer_height = {final_leftover_for_first_layer}")
        first_layer_height = final_leftover_for_first_layer
    
    print(f"\n=== 最終結果 ===")
    print(f"総段数: {num_stages}")
    print(f"1層目高さ: {first_layer_height}mm")
    print(f"ジャッキアップ高さ: {jack_up_height}mm")
    print(f"コマ数: {modules_count}")
    print(f"根がらみ支柱設置: {'可能' if tie_possible else '不可'}")
    print(f"reduction_loops: {reduction_loops}")
    
    return {
        "num_stages": num_stages,
        "first_layer_height": first_layer_height,
        "jack_up_height": jack_up_height,
        "modules_count": modules_count,
        "tie_possible": tie_possible
    }

# テスト実行
debug_height_calculation(6250, "勾配軒", True, 3)

File: debug-height-calc.js
// 高さ計算ロジックの詳細デバッグ
const { calculateAll } = require('./packages/core/dist/calculator/engine');

console.log('=== 高さ計算ロジックの詳細デバッグ ===\n');

// 屋根形状別の基本高さマッピング
const ROOF_BASE_UNIT_MAP = {
  'フラット': 1700,
  '勾配軒': 1900,
  '陸屋根': 1800,
};

// 定数値
const STAGE_UNIT_HEIGHT = 1900;
const FIRST_LAYER_MIN_HEIGHT_THRESHOLD = 950;
const TIE_COLUMN_REDUCTION_LARGE = 475;
const TIE_COLUMN_REDUCTION_SMALL = 130;
const TIE_COLUMN_MIN_HEIGHT_FOR_LARGE_REDUCTION = 550;
const TIE_COLUMN_MIN_HEIGHT_FOR_SMALL_REDUCTION = 150;

function debugHeightCalculation(input) {
  console.log(`\n--- ${input.roof_shape} 高さ${input.standard_height}mm の詳細計算 ---`);
  
  const baseUnit = ROOF_BASE_UNIT_MAP[input.roof_shape] || 1700;
  console.log(`1. 基本高さ (${input.roof_shape}): ${baseUnit}mm`);
  
  const remainder = input.standard_height - baseUnit;
  console.log(`2. 残り高さ: ${input.standard_height} - ${baseUnit} = ${remainder}mm`);
  
  const stageUnit = STAGE_UNIT_HEIGHT;
  console.log(`3. 段数単位高さ: ${stageUnit}mm`);
  
  const initialStages = 1 + (remainder > 0 ? Math.floor(remainder / stageUnit) : 0);
  console.log(`4. 初期段数: 1 + floor(${remainder} / ${stageUnit}) = ${initialStages}`);
  
  const initialLeftover = remainder - (initialStages - 1) * stageUnit;
  console.log(`5. 初期余り: ${remainder} - (${initialStages} - 1) * ${stageUnit} = ${initialLeftover}mm`);
  
  let firstLayerHeight;
  if (initialLeftover < FIRST_LAYER_MIN_HEIGHT_THRESHOLD) {
    firstLayerHeight = initialLeftover + stageUnit;
    console.log(`6. 1層目高さ (余り < ${FIRST_LAYER_MIN_HEIGHT_THRESHOLD}): ${initialLeftover} + ${stageUnit} = ${firstLayerHeight}mm`);
  } else {
    firstLayerHeight = initialLeftover;
    console.log(`6. 1層目高さ (余り >= ${FIRST_LAYER_MIN_HEIGHT_THRESHOLD}): ${firstLayerHeight}mm`);
  }
  
  const remainingAfterFirst = input.standard_height - baseUnit - firstLayerHeight;
  console.log(`7. 1層目後の残り: ${input.standard_height} - ${baseUnit} - ${firstLayerHeight} = ${remainingAfterFirst}mm`);
  
  const numStages = 1 + (remainingAfterFirst > 0 ? Math.floor(remainingAfterFirst / stageUnit) : 0);
  console.log(`8. 最終段数: 1 + floor(${remainingAfterFirst} / ${stageUnit}) = ${numStages}`);
  
  const leftover = input.standard_height - baseUnit - (numStages - 1) * stageUnit;
  console.log(`9. 最終余り: ${input.standard_height} - ${baseUnit} - (${numStages} - 1) * ${stageUnit} = ${leftover}mm`);
  
  let jackUpHeight = leftover;
  let reductionLoops = 0;
  let tiePossible = true;
  
  console.log(`10. 根がらみ支柱計算開始 (tie_column: ${input.tie_column})`);
  
  if (input.tie_column) {
    console.log(`    初期ジャッキアップ高さ: ${jackUpHeight}mm`);
    
    if (jackUpHeight >= TIE_COLUMN_MIN_HEIGHT_FOR_LARGE_REDUCTION) {
      console.log(`    大減算適用可能 (>= ${TIE_COLUMN_MIN_HEIGHT_FOR_LARGE_REDUCTION}mm)`);
      while (jackUpHeight >= TIE_COLUMN_MIN_HEIGHT_FOR_LARGE_REDUCTION) {
        jackUpHeight -= TIE_COLUMN_REDUCTION_LARGE;
        reductionLoops++;
        console.log(`    大減算 ${reductionLoops}回目: ${jackUpHeight}mm (${TIE_COLUMN_REDUCTION_LARGE}mm減算)`);
      }
      
      if (jackUpHeight >= TIE_COLUMN_MIN_HEIGHT_FOR_SMALL_REDUCTION) {
        jackUpHeight -= TIE_COLUMN_REDUCTION_SMALL;
        console.log(`    小減算適用: ${jackUpHeight}mm (${TIE_COLUMN_REDUCTION_SMALL}mm減算)`);
      } else {
        tiePossible = false;
        jackUpHeight = leftover;
        reductionLoops = 0;
        console.log(`    小減算不可 (< ${TIE_COLUMN_MIN_HEIGHT_FOR_SMALL_REDUCTION}mm) - 根がらみ支柱NG`);
      }
    } else if (jackUpHeight >= TIE_COLUMN_MIN_HEIGHT_FOR_SMALL_REDUCTION) {
      jackUpHeight -= TIE_COLUMN_REDUCTION_SMALL;
      console.log(`    小減算のみ適用: ${jackUpHeight}mm (${TIE_COLUMN_REDUCTION_SMALL}mm減算)`);
    } else {
      tiePossible = false;
      jackUpHeight = leftover;
      console.log(`    減算不可 (< ${TIE_COLUMN_MIN_HEIGHT_FOR_SMALL_REDUCTION}mm) - 根がらみ支柱NG`);
    }
  } else {
    console.log(`    根がらみ支柱未使用 - 通常減算のみ`);
    while (jackUpHeight >= TIE_COLUMN_REDUCTION_LARGE) {
      jackUpHeight -= TIE_COLUMN_REDUCTION_LARGE;
      reductionLoops++;
      console.log(`    通常減算 ${reductionLoops}回目: ${jackUpHeight}mm (${TIE_COLUMN_REDUCTION_LARGE}mm減算)`);
    }
  }
  
  console.log(`11. 最終ジャッキアップ高さ: ${jackUpHeight}mm`);
  console.log(`12. 減算ループ回数: ${reductionLoops}`);
  console.log(`13. 根がらみ支柱可能: ${tiePossible}`);
  
  // コマ数計算
  let modulesCount = 4 + (numStages - 1) * 4 + reductionLoops;
  if (input.railing_count === 3) {
    modulesCount += 2;
  } else if (input.railing_count === 2) {
    modulesCount += 1;
  }
  console.log(`14. コマ数: 4 + (${numStages} - 1) * 4 + ${reductionLoops} + 手すり補正 = ${modulesCount}`);
  
  // 1層目高さの最終調整
  const finalLeftoverForFirstLayer = input.standard_height - baseUnit - (numStages - 1) * stageUnit;
  let finalFirstLayerHeight;
  if (finalLeftoverForFirstLayer < FIRST_LAYER_MIN_HEIGHT_THRESHOLD) {
    finalFirstLayerHeight = finalLeftoverForFirstLayer + stageUnit;
    console.log(`15. 1層目最終高さ (< ${FIRST_LAYER_MIN_HEIGHT_THRESHOLD}): ${finalLeftoverForFirstLayer} + ${stageUnit} = ${finalFirstLayerHeight}mm`);
  } else {
    finalFirstLayerHeight = finalLeftoverForFirstLayer;
    console.log(`15. 1層目最終高さ (>= ${FIRST_LAYER_MIN_HEIGHT_THRESHOLD}): ${finalFirstLayerHeight}mm`);
  }
  
  console.log('\n予想結果:');
  console.log(`  段数: ${numStages}`);
  console.log(`  1層目高さ: ${finalFirstLayerHeight}mm`);
  console.log(`  ジャッキアップ高さ: ${jackUpHeight}mm`);
  console.log(`  根がらみ支柱OK: ${tiePossible}`);
  console.log(`  コマ数: ${modulesCount}`);
  
  // 実際の計算実行
  console.log('\n実際の計算結果:');
  const result = calculateAll(input);
  console.log(`  段数: ${result.num_stages}`);
  console.log(`  1層目高さ: ${result.first_layer_height}mm`);
  console.log(`  ジャッキアップ高さ: ${result.jack_up_height}mm`);
  console.log(`  根がらみ支柱OK: ${result.tie_ok}`);
  console.log(`  コマ数: ${result.modules_count}`);
  
  // 矛盾チェック
  console.log('\n矛盾チェック:');
  if (numStages !== result.num_stages) {
    console.log(`❌ 段数が一致しません: 予想${numStages} vs 実際${result.num_stages}`);
  }
  if (Math.abs(finalFirstLayerHeight - result.first_layer_height) > 1) {
    console.log(`❌ 1層目高さが一致しません: 予想${finalFirstLayerHeight} vs 実際${result.first_layer_height}`);
  }
  if (Math.abs(jackUpHeight - result.jack_up_height) > 1) {
    console.log(`❌ ジャッキアップ高さが一致しません: 予想${jackUpHeight} vs 実際${result.jack_up_height}`);
  }
  if (tiePossible !== result.tie_ok) {
    console.log(`❌ 根がらみ支柱判定が一致しません: 予想${tiePossible} vs 実際${result.tie_ok}`);
  }
  if (modulesCount !== result.modules_count) {
    console.log(`❌ コマ数が一致しません: 予想${modulesCount} vs 実際${result.modules_count}`);
  }
  
  // 高さ整合性チェック
  const calculatedTotal = baseUnit + result.first_layer_height + (result.num_stages - 1) * stageUnit;
  if (Math.abs(calculatedTotal - input.standard_height) > 1) {
    console.log(`❌ 高さ整合性エラー: ${baseUnit} + ${result.first_layer_height} + (${result.num_stages} - 1) * ${stageUnit} = ${calculatedTotal} ≠ ${input.standard_height}`);
  } else {
    console.log(`✅ 高さ整合性OK: ${baseUnit} + ${result.first_layer_height} + (${result.num_stages} - 1) * ${stageUnit} = ${calculatedTotal} = ${input.standard_height}`);
  }
}

// テストケース実行
const testCases = [
  {
    name: '基本ケース',
    data: {
      width_NS: 10000, width_EW: 8000,
      eaves_N: 500, eaves_E: 500, eaves_S: 500, eaves_W: 500,
      boundary_N: null, boundary_E: null, boundary_S: null, boundary_W: null,
      standard_height: 6000, roof_shape: 'フラット',
      tie_column: true, railing_count: 2,
      use_355_NS: 0, use_300_NS: 0, use_150_NS: 0,
      use_355_EW: 0, use_300_EW: 0, use_150_EW: 0,
      target_margin_N: 900, target_margin_E: 900, target_margin_S: 900, target_margin_W: 900
    }
  },
  {
    name: '高い建物',
    data: {
      width_NS: 10000, width_EW: 8000,
      eaves_N: 500, eaves_E: 500, eaves_S: 500, eaves_W: 500,
      boundary_N: null, boundary_E: null, boundary_S: null, boundary_W: null,
      standard_height: 12000, roof_shape: '勾配軒',
      tie_column: true, railing_count: 2,
      use_355_NS: 0, use_300_NS: 0, use_150_NS: 0,
      use_355_EW: 0, use_300_EW: 0, use_150_EW: 0,
      target_margin_N: 900, target_margin_E: 900, target_margin_S: 900, target_margin_W: 900
    }
  },
  {
    name: '問題ケース - 境界値',
    data: {
      width_NS: 10000, width_EW: 8000,
      eaves_N: 500, eaves_E: 500, eaves_S: 500, eaves_W: 500,
      boundary_N: null, boundary_E: null, boundary_S: null, boundary_W: null,
      standard_height: 4850, roof_shape: 'フラット',
      tie_column: true, railing_count: 2,
      use_355_NS: 0, use_300_NS: 0, use_150_NS: 0,
      use_355_EW: 0, use_300_EW: 0, use_150_EW: 0,
      target_margin_N: 900, target_margin_E: 900, target_margin_S: 900, target_margin_W: 900
    }
  }
];

testCases.forEach(testCase => {
  console.log(`\n${'='.repeat(60)}`);
  console.log(`テストケース: ${testCase.name}`);
  console.log(`${'='.repeat(60)}`);
  debugHeightCalculation(testCase.data);
});

console.log('\n=== デバッグ完了 ===');